File: solutions/0204. Count Primes/0204.cpp
// Time complexity: O(nlog(logn))
// Space complexity: O(n)
class Solution {
public:
    int countPrimes(int n) {
        if ( n <= 2) {
            return 0;
        }

        vector<bool> isPrime(n, true);
        isPrime[0] = false;
        isPrime[1] = false;
        for (int p = 2; p * p < n; ++p) {
            if (isPrime[p]) {
                for (int i = p * p; i < n; i += p) {
                    isPrime[i] = false;
                }
            }
        }
        int count = 0;
        for (int i = 2; i < n; ++i) {
            if (isPrime[i]) {
                count++;
            }
        }
        return count;
    }
};

File: solutions/0204. Count Primes/0204.py
# Time complexity: O(nlog(logn))
# Space complexity: O(n)
class Solution:
  def countPrimes(self, n: int) -> int:
    if n <= 2:
      return 0
        
    isPrime = [True] * n
    isPrime[0] = False
    isPrime[1] = False
        
    p = 2
    while p * p < n:
      if isPrime[p]:
        for i in range(p * p, n, p):
          isPrime[i] = False
      p += 1
        
    count = 0
    for i in range(2, n):
      if isPrime[i]:
        count += 1
        
    return count       